tests/testthat/helper-shinytest2.R

```diff
@@ -0,0 +1,223 @@
+library(shinytest2)
+library(glue)
+library(rvest)
+
+module_ns_shiny2 <- function(app) {
+  source <- app$get_html("html", outer_html = TRUE)
+  module_id <- rvest::html_attr(
+    rvest::html_node(rvest::read_html(source), css = ".teal_module"),
+    "id"
+  )
+  NS(paste0(module_id, "-module"))
+}
+
+global_fns <- NS("teal-main_ui-filter_panel")
+default_idle_timeout <- 20000
+
+#' Get the `AppDriver` to test the app created from `init`
+#'
+#' Create the `AppDriver` using the `init` parameters of `teal` to help with testing.
+#'
+#' @inheritParams init
+#' @keywords internal
+#' @return  (`AppDriver`) object which can be used to test the app using `shinytest2`.
+get_test_app_object <- function(data,
+                                modules,
+                                filter = teal_slices(),
+                                title = build_app_title(),
+                                header = tags$p(),
+                                footer = tags$p(),
+                                id = character(0)) {
+  app <- init(
+    data = data,
+    modules = modules,
+    filter = filter,
+    title = title,
+    header = header,
+    footer = footer,
+    id = id
+  )
+  suppressWarnings(
+    AppDriver$new(
+      shinyApp(app$ui, app$server),
+      name = "teal",
+      variant = platform_variant(),
+      load_timeout = 300000,
+      seed = 123
+    )
+  )
+}
```

Comment:
I just had an issue when running the same test repeatedly in a short period.

What do you think about stopping the app automatically just like `{withr}` (maybe the name could also change to follow convention.

```suggestion
local_test_app_object <- function(data,
                                  modules,
                                  filter = teal_slices(),
                                  title = build_app_title(),
                                  header = tags$p(),
                                  footer = tags$p(),
                                  id = character(0),
                                  envir = parent.frame()) {
  app <- init(
    data = data,
    modules = modules,
    filter = filter,
    title = title,
    header = header,
    footer = footer,
    id = id
  )
  test_app <- suppressWarnings(
    shinytest2::AppDriver$new(
      shiny::shinyApp(app$ui, app$server),
      name = "teal",
      variant = shinytest2::platform_variant(),
      load_timeout = 300000,
      seed = 123
    )
  )

  if (requireNamespace("withr", quietly = TRUE)) {
    withr::defer(test_app$stop(), envir = envir)
  }

  test_app
}
```

Reply:
Interesting. For me, the port crash never occurred. Ideally, I would term this as a bug in `shinytest2`. Let me try to recreate this.

Reply:
I'm resolving this as tests now have a `app$stop()` call.

Maybe we could still use this strategy or add a `withr::defer(app$stop())` to make sure that the app stops cleanly even if there are errors.